### PR TITLE
Re-use ResolveFieldContextAdapter<T> instances

### PR DESCRIPTION
**Pull request by Shane32**
This PR is an implementation to reuse `ResolveFieldContextAdapter<T>` instances created by `FuncFieldResolver<TSourceType, TReturnType>` whenever possible, as already described by the FAQ for `ReadonlyResolveFieldContext` instances.

In order to optimize for execution speed, upon creating a `FuncFieldResolver<,>`, a number of checks are run to determine the most optimal execution path, and then the resolver is built and stored.

The introspection benchmark makes heavy use of `NameFieldResolver` and `ExpressionFieldResolver` which is unaffected by these changes.  Even so, with the few uses of `FuncFieldResolver`, the benchmarks demonstrate a 2% decrease in memory and 2% faster execution speed.  The hero sample does not use a typed `FuncFieldResolver` so there is no change there.

For implementations which rely heavily on `FuncFieldResolver<TSourceType, TReturnType>`, there is likely to a be a much greater amount of benefit.

BEFORE

|        Method | UseCaching |       Mean |     Error |    StdDev |   Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |----------- |-----------:|----------:|----------:|--------:|--------:|------:|----------:|
| Introspection |      False | 960.426 us | 5.2523 us | 4.9130 us | 78.1250 | 23.4375 |     - | 487.05 KB |
|          Hero |      False |  16.958 us | 0.0787 us | 0.0698 us |  1.4038 |       - |     - |   8.73 KB |
| Introspection |       True | 722.193 us | 5.4136 us | 5.0639 us | 72.2656 | 22.4609 |     - | 447.35 KB |
|          Hero |       True |   3.530 us | 0.0144 us | 0.0135 us |  0.6599 |  0.0038 |     - |   4.05 KB |

AFTER

|        Method | UseCaching |       Mean |     Error |    StdDev |   Gen 0 |   Gen 1 | Gen 2 | Allocated |
|-------------- |----------- |-----------:|----------:|----------:|--------:|--------:|------:|----------:|
| Introspection |      False | 946.462 us | 4.7088 us | 4.4046 us | 78.1250 | 30.2734 |     - | 478.64 KB |
|          Hero |      False |  16.927 us | 0.1174 us | 0.1098 us |  1.4038 |       - |     - |   8.73 KB |
| Introspection |       True | 722.684 us | 4.4643 us | 3.9575 us | 71.2891 | 21.4844 |     - | 438.95 KB |
|          Hero |       True |   3.583 us | 0.0140 us | 0.0124 us |  0.6599 |  0.0038 |     - |   4.05 KB |

### CODE REVIEW

File: src/GraphQL/Resolvers/FuncFieldResolver.cs [hunk 1]
```diff
@@ -30,16 +32,44 @@ public FuncFieldResolver(Func<IResolveFieldContext, TReturnType> resolver)
...
+                    if (adapter == null)
+                    {
+                        adapter = new ResolveFieldContextAdapter<TSourceType>(context);
+                    }
+                    else
+                    {
+                        adapter.Set(context);
+                    }
+                    var ret = resolver(adapter);
+                    if (!(ret is Task task && !task.IsCompleted) && !(ret is IDataLoaderResult))
```
Comment (Shane32) [inline src/GraphQL/Resolvers/FuncFieldResolver.cs:61]:
```suggestion
                    // only re-use contexts that completed synchronously
                    if (!(ret is Task task && !task.IsCompleted) && !(ret is IDataLoaderResult))
```

File: src/GraphQL.Benchmarks/Properties/launchSettings.json [hunk 2]
```diff
@@ -1,8 +1,7 @@
 {
   "profiles": {
     "Profiler": {
-      "commandName": "Project",
-      "commandLineArgs": "test"
+      "commandName": "Project"
```
Comment (sungam3r) [inline src/GraphQL.Benchmarks/Properties/launchSettings.json:4]:
```suggestion
      "commandName": "Project",
      "commandLineArgs": "test"
```

The point is to run the profiler, not benchmarks. If you wish then add second profile like "Bench" without args.

**Comment by sungam3r** [comment]
I don't think it's worth doing that. The gain is minimal compared to the complexity of the code. Basically, we should apply optimizations from #2166 , namely, consider the possibility of abandoning `ResolveFieldContext`/`ResolveFieldContextAdapter`.

**Comment by Shane32** [comment]
It's only minimal because the introspection query rarely uses FuncFieldResolver.  I think in real-world scenarios the gain will be much greater.

Unfortunately it is even more complex than the current sample.

**Comment by Shane32** [comment]
I have a very large graphql application running, and there is very little, if any, use of `ExpressionFieldResolver`, and no use of `NameFieldResolver`.  Applying internal optimizations to those specific classes would be of little benefit to me.  Most of my resolvers are custom implementations of `IFieldResolver` and/or `FuncFieldResolver<TSource>`.  I also have extensive code that takes advantage of the fact that `IResolveFieldContext` is an interface and can be extended.  Completely doing away with `IResolveFieldContext` isn't an option.  Providing optimizations to `IResolveFieldContext` and `IResolveFieldContext<TSource>` can definitely improve performance for my use.  Just because the introspection/hero tests don't use `FuncFieldResolver` much doesn't mean that this isn't a very valuable change for some people like me.

**Comment by Shane32** [comment]
Simplification isn't possible since the typical field builders don't strongly type `TReturnType` 😢

**Comment by sungam3r** [comment]
>  Just because the introspection/hero tests don't use FuncFieldResolver much doesn't mean that this isn't a very valuable change for some people like me.

Let's add another specific bench.

**Comment by sungam3r** [comment]
>  Completely doing away with IResolveFieldContext isn't an option. 

Not completely but doing something that avoids this "intermediate" object.

**Comment by sungam3r** [comment]
Describe your main idea. What is the code intended to do?

**Comment by Shane32** [comment]
> Describe your main idea. What is the code intended to do?

This code is intended to reuse `IResolveFieldContext<TSourceType>` instances created by `FuncFieldResolver<TSourceType, TReturnType>` whenever possible, as already described by the FAQ for `IResolveFieldContext` instances.

**Comment by Shane32** [comment]
@sungam3r I suggest you review this code.  I don't see any reason not to merge it.  It clearly provides a measurable impact for the better.  Many other performance changes we have made have provided smaller benefits, and at 100 lines of code, this is not so very complicated of a change.  There are no public API changes.

**Comment by sungam3r** [comment]
I would like to get rid of `ResolveFieldContextAdapter` at all and simplify code.

> For implementations which rely heavily on FuncFieldResolver<TSourceType, TReturnType>, there is likely to a be a much greater amount of benefit.

Let's prove it.

**Comment by Shane32** [comment]
> I would like to get rid of `ResolveFieldContextAdapter` at all and simplify code.

When that happens, we can delete this code. Not a reason not to merge it now.

**Comment by sungam3r** [comment]
I'm reviewing right now.

**Comment by sungam3r** [comment]
I would not to maintain code like this. It has become more difficult to perceive with all these `if/else` branches. The winnings are minimal. Most likely, `ResolveFieldContextAdapter` will be removed soon and there will be no point in saving on it. I have not yet found the time to apply optimizations from my other branch. So... if you want, merge it. I hope that I have time to change this code before the v4 release.

**Comment by Shane32** [comment]
> So... if you want, merge it. I hope that I have time to change this code before the v4 release.

I added a lot of tests to prove functionality.  I'm ready to merge this now.

**Comment by sungam3r** [comment]
OK. I don't know how long this code will stay. Maybe forever. It is difficult for me to understand it and it will be difficult to maintain it.

**Pull request merged**